### PR TITLE
Add `conda index` plugin

### LINKED ISSUE #81: Provide a conda subcommand plugin
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

Conda has a new plugin system through which we can provide "conda index" (without using entry_points scripts). Provide one.

Will still need to figure out what to do about conda-build's "conda-index" script, which can also be called as "conda index".

Our command line arguments are necessarily different than legacy conda-index.

### Why is this needed?

_No response_

### What should happen?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dholth**
### Description

Fix #81 

Note that this has a different CLI than the old "conda-index" command and would generally break conda-index scripts. This uses a click CLI and is more explicit. For example it won't automatically index the current directory if passed no arguments. For that reason we have been recommending `python -m conda_index` to use the many-times-faster standalone conda-index.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [ ] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

File: conda_index/plugin.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+"""
+Conda command plugin, provide "conda reindex" command, renamed to
+avoid clash with old "conda index" CLI.
+"""
+
+import conda.plugins
+
+def command(*args):
+     import conda_index.cli
+     return conda_index.cli.cli()
```
Comment (jezdez) [inline conda_index/plugin.py:10]:
```suggestion
     return conda_index.cli.cli(prog_name="conda reindex", args=args)
```

Comment (jezdez) [reply conda_index/plugin.py:10]: That should suffice to tell click about the passed options

Comment (dholth) [reply conda_index/plugin.py]: It looks like there is an unexpected tuple with this change
```
% ~/miniconda3/envs/conda-index-foo/bin/conda reindex --help

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/Users/dholth/miniconda3/envs/conda-index-foo/lib/python3.10/site-packages/conda/exception_handler.py", line 17, in __call__
        return func(*args, **kwargs)
      File "/Users/dholth/miniconda3/envs/conda-index-foo/lib/python3.10/site-packages/conda/cli/main.py", line 78, in main_subshell
        exit_code = do_call(args, parser)
      File "/Users/dholth/miniconda3/envs/conda-index-foo/lib/python3.10/site-packages/conda/cli/conda_argparse.py", line 142, in do_call
        result = plugin_subcommand.action(getattr(args, "_args", args))
      File "/Users/dholth/prog/conda-index/conda_index/plugin.py", line 10, in command
        return conda_index.cli.cli(prog_name="conda reindex", args=args)
      File "/Users/dholth/miniconda3/envs/conda-index-foo/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
        return self.main(*args, **kwargs)
      File "/Users/dholth/miniconda3/envs/conda-index-foo/lib/python3.10/site-packages/click/core.py", line 1053, in main
        rv = self.invoke(ctx)
      File "/Users/dholth/miniconda3/envs/conda-index-foo/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/Users/dholth/miniconda3/envs/conda-index-foo/lib/python3.10/site-packages/click/core.py", line 754, in invoke
        return __callback(*args, **kwargs)
      File "/Users/dholth/prog/conda-index/conda_index/cli/__init__.py", line 134, in cli
        channel_index.index(
      File "/Users/dholth/prog/conda-index/conda_index/index/__init__.py", line 542, in index
        subdirs = self.detect_subdirs()
      File "/Users/dholth/prog/conda-index/conda_index/index/__init__.py", line 710, in detect_subdirs
        for subdir in os.scandir(self.channel_root)
    FileNotFoundError: [Errno 2] No such file or directory: "/Users/dholth/prog/conda-index/('--help',)"
```

File: conda_index/plugin.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+"""
+Conda command plugin, provide "conda reindex" command, renamed to
+avoid clash with old "conda index" CLI.
+"""
+
+import conda.plugins
+
+def command(*args):
```
Comment (kenodegard) [inline conda_index/plugin.py:8]:
```suggestion
def command(args):
```

File: conda_index/plugin.py (hunk 1 above)
Comment (jezdez) [reply conda_index/plugin.py]: Fixed by https://github.com/conda/conda-index/pull/131/commits/78c67453a5753fe2ebc193ff091d65a95911a169

**Review by jezdez — APPROVED** [review]

**Comment by dholth** [comment]
LMK if you have a chance to test this, I'm not sure if it works correctly or not.

**Review by jezdez — APPROVED** [review]
I think this is a good idea!

**Comment by dholth** [comment]
This interacts badly with the click parser inside conda_index

File: pyproject.toml [hunk 3]
```diff
@@ -42,6 +42,9 @@ docs = [
 # conda-build also provides conda-index script.
 # require python -m conda_index for the moment to avoid the conflict.
 
+[project.entry-points."conda"]
+reindex = "conda_index.plugin"
```
Comment (kenodegard) [inline pyproject.toml:46]:
```suggestion
[project.entry-points.conda]
index = "conda_index.plugin"
```

File: conda_index/plugin.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+def command(args):
+     import conda_index.cli
+     return conda_index.cli.cli(prog_name="conda reindex", args=args)
+
+@conda.plugins.hookimpl
+def conda_subcommands():
+    yield conda.plugins.CondaSubcommand(
+        name="reindex",
+        action=command,
+        summary="Update package index metadata files.  Replaces `conda index`."
```
Comment (kenodegard) [inline conda_index/plugin.py:17]:
```suggestion
        name="index",
        action=command,
        summary="Update package index metadata files."
```

**Review by kenodegard — CHANGES_REQUESTED** [review]
We should move ahead and simply remove the deprecated `conda index` subcommand in `conda-build 24.1.0`.

**Comment by dholth** [comment]
I tried this command today, and it seemed to work correctly. (Previously there were issues with the subcommand name being interpreted as one of the arguments to the parser).

**Review by jezdez — APPROVED** [review]

**Review by kenodegard — APPROVED** [review]

**Comment by kenodegard** [comment]
we should also add `conda index` to the recipe's `test/commands`

File: conda_index/plugin.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import conda.plugins
+
+def command(args):
+     import conda_index.cli
+     return conda_index.cli.cli(prog_name="conda index", args=args)
+
+
+@conda.plugins.hookimpl
+def conda_subcommands():
+    # only expose plugin if conda-build>=24.1.0
```
Comment (kenodegard) [inline conda_index/plugin.py:14]:
```suggestion
    # hide plugin if conda-build < 24.1.0
```

**Pull request merged**